### PR TITLE
[Performance] Replace md5 hash function by crc32 on subscriptions table

### PR DESCRIPTION
**Pull request by nickhsine**
@babygoat

### CODE REVIEW

File: storage/subscription.go [hunk 1]
```diff
@@ -1,26 +1,54 @@
...
+
+	// find if endpoint is existed or not
+	if err = tx.Find(&ws, "hash_endpoint = ? AND endpoint = ?", wpSub.HashEndpoint, wpSub.Endpoint).Error; err != nil {
+		// endpoint is not in the DB
+		if IsRecordNotFoundError(err) {
+			// create the subscription
+			err = tx.Create(&wpSub).Error
+			// no error occurs, commit the transaction
+			if err == nil {
+				return tx.Commit().Error
```
Comment (babygoat) [inline storage/subscription.go]: What if an error occurs during the transaction commit? Is it possible for controller to handle the error?

Comment (nickhsine) [reply storage/subscription.go]: The error will passed to controller, and controller will return internal server error.

File: storage/errors.go [hunk 2]
```diff
@@ -24,14 +24,26 @@ var ErrDuplicateEntry uint16 = 1062
...
+	errStruct, ok := err.(*mysql.MySQLError)
+	if ok && errStruct.Number == ErrDuplicateEntry {
+		return true
+	}
+	return false
+}
+
 // NewStorageError return AppError with detailed information.
 // This method is mainly used to deal with MySQL error.
 func (g *GormStorage) NewStorageError(err error, where string, message string) error {
```
Comment (babygoat) [inline storage/errors.go]: When it comes to deal with the if-else-chain, I suggest use switch instead.

switch {
     case IsRecordNotFoundError(err):
              stmt1
     case IsDuplicateEntryError(err):
              stmt2
     default:
              default handling
}

File: storage/bookmark.go [hunk 3]
```diff
@@ -83,7 +83,7 @@ func (g *GormStorage) GetBookmarksOfAUser(id string, limit, offset int) ([]model
 
 // CreateABookmarkOfAUser this func will create a bookmark and build the relationship between the bookmark and the user
 func (g *GormStorage) CreateABookmarkOfAUser(userID string, bookmark models.Bookmark) (models.Bookmark, error) {
-	var _bookmark models.Bookmark = bookmark
+	var _bookmark = bookmark
 	var err error
 	var user models.User
```
Comment (babygoat) [inline storage/bookmark.go]: Should below update operation wrapping within a transaction group?

Comment (nickhsine) [reply storage/bookmark.go]: Yeah, it should be wrapped in a transaction.
I will fix it in the next PR.

File: membership_user.sql [hunk 4]
```diff
@@ -183,13 +183,13 @@ DROP TABLE IF EXISTS `web_push_subscriptions`;
...
-  `hash_endpoint` char(32) NOT NULL,
+  `hash_endpoint` int(10) unsigned NOT NULL,
   `keys` varchar(200) NOT NULL,
   `expiration_time` timestamp NULL DEFAULT NULL,
   `user_id` int(10) unsigned NULL DEFAULT NULL,
   `created_at` timestamp DEFAULT CURRENT_TIMESTAMP,
   `updated_at` timestamp DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
   PRIMARY KEY (`id`),
-  UNIQUE KEY `uix_web_push_subscriptions_hash_endpoint` (`hash_endpoint`)
+  KEY `uix_web_push_subscriptions_hash_endpoint` (`hash_endpoint`)
```
Comment (babygoat) [inline membership_user.sql]: Minor. If the index is no longer unique, I suggest the naming should be adapted accordingly.

**Comment by babygoat** [comment]
Except for the minor suggestion, others look good to me now.

File: membership_user.sql (hunk 4 above)
Comment (nickhsine) [reply membership_user.sql]: I will update this key name before merging this PR.

**Comment by babygoat** [comment]
LGTM

File: models/subscription.go [hunk 5]
```diff
@@ -29,3 +29,8 @@ func (wpSub *WebPushSubscription) SetExpirationTime(expireTime int64) {
 func (wpSub *WebPushSubscription) SetUserID(userID uint) {
 	wpSub.UserID = &userID
 }
+
+// set WebPushSubscription's table name ti be `web_push_subs`
```
Comment (babygoat) [inline models/subscription.go]: typo to?

**Pull request merged**